src/composites.rs

```diff
@@ -0,0 +1,214 @@
+/// generated by the CompositeNewService will pass a request to the first service
+/// in the list whose base path is a prefix of the request's path or return a
+/// "not found" response if there is no match.
+pub struct CompositeNewService<U, V, W>(Vec<(&'static str, Box<BoxedNewService<U, V, W>>)>)
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static;
+
+impl<U: GetPath, V: NotFound, W> CompositeNewService<U, V, W> {
+    /// Create an empty CompositeNewService
+    pub fn new() -> Self {
+        CompositeNewService(Vec::new())
+    }
+
+    /// Add a new NewService with a base path to the composite
+    pub fn append_new_service(
+        &mut self,
+        base_path: &'static str,
+        new_service: Box<BoxedNewService<U, V, W>>,
+    ) {
+        self.0.push((base_path, new_service));
+    }
+}
+
+impl<U, V, W> NewService for CompositeNewService<U, V, W>
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static,
+{
+    type Request = U;
+    type Response = V;
+    type Error = W;
+    type Instance = CompositeService<U, V, W>;
+
+    fn new_service(&self) -> Result<Self::Instance, io::Error> {
+        let mut vec = Vec::new();
+
+        for &(base_path, ref new_service) in self.0.iter() {
+            vec.push((base_path, new_service.boxed_new_service()?))
+        }
+
+        Ok(CompositeService(vec))
+    }
+}
+
+impl<U, V, W> fmt::Debug for CompositeNewService<U, V, W>
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        // Get vector of base paths
+        let str_vec: Vec<&'static str> = self.0.iter().map(|&(base_path, _)| base_path).collect();
+        write!(
+            f,
+            "CompositeNewService accepting base paths: {:?}",
+            str_vec,
+        )
+    }
+}
+
+/// A struct combining multiple hyper Services
+pub struct CompositeService<U, V, W>(
+    Vec<
+        (&'static str,
+         Box<
+            Service<
+                Request = U,
+                Response = V,
+                Error = W,
+                Future = Box<Future<Item = V, Error = W>>,
+            >,
+        >),
+    >
+)
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static;
+
+impl<U, V, W> Service for CompositeService<U, V, W>
+where
+    U: GetPath,
+    V: NotFound + 'static,
+    W: 'static,
+{
+    type Request = U;
+    type Response = V;
+    type Error = W;
+    type Future = Box<Future<Item = V, Error = W>>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+
+        let mut result = None;
+
+        for &(base_path, ref service) in self.0.iter() {
+            if req.path().starts_with(base_path) {
+                result = Some(service.call(req));
+                break;
+            }
+        }
+
+        if let Some(result) = result {
+            result
+        } else {
+            Box::new(future::ok(V::not_found()))
+        }
```

Comment:
@tca-ms I think these 5 lines are equivalent to `result.unwrap_or(Box::new(future::ok(V::not_found())))` (https://doc.rust-lang.org/core/option/enum.Option.html#method.unwrap_or)

Reply:
👍